File: openHAB/OpenHABTracker.m
//
//  OpenHABTracker.m
//  openHAB
//
//  Created by Victor Belov on 13/01/14.
//  Copyright (c) 2014 Victor Belov. All rights reserved.
//

#import "OpenHABTracker.h"
#import <sys/socket.h>
#import <netinet/in.h>
#import <arpa/inet.h>
#import <netdb.h>
#import <SystemConfiguration/SCNetworkReachability.h>
#import "Reachability+URL.h"
#import <FastSocket.h>

@implementation OpenHABTracker
@synthesize openHABDemoMode, openHABLocalUrl, openHABRemoteUrl, delegate, netService, reach;

- (OpenHABTracker *)init
{
    self = [super init];
    NSUserDefaults *prefs = [NSUserDefaults standardUserDefaults];
    openHABDemoMode = [prefs boolForKey:@"demomode"];
    openHABLocalUrl = [prefs valueForKey:@"localUrl"];
    openHABRemoteUrl = [prefs valueForKey:@"remoteUrl"];
    return self;
}

- (void)startTracker
{
    // Check if any network is available
    if ([self isNetworkConnected]) {
        // Check if demo mode is switched on in preferences
        if (self.openHABDemoMode) {
            NSLog(@"OpenHABTracker demo mode preference is on");
            [self trackedDemoMode];
        } else {
            // Check if network is WiFi. If not, go for remote URL
            if (![self isNetworkWiFi]) {
                NSLog(@"OpenHABTracker network is not WiFi");
                [self trackedRemoteUrl];
            // If it is WiFi
            } else {
                NSLog(@"OpenHABTracker network is Wifi");
                // Check if local URL is configured, if yes
                if ([openHABLocalUrl length] > 0) {
                    if ([self isURLReachable:[NSURL URLWithString:openHABLocalUrl]]) {
                        [self trackedLocalUrl];
                    } else {
                        [self trackedRemoteUrl];
                    }
                // If not, go for Bonjour discovery
                } else {
                    [self startDiscovery];
                }
            }
        }
    } else {
        if (self.delegate && [self.delegate respondsToSelector:@selector(openHABTrackingError:)]) {
            NSMutableDictionary *errorDetail = [NSMutableDictionary dictionary];
            [errorDetail setValue:@"Network is not available." forKey:NSLocalizedDescriptionKey];
            NSError *trackingError = [NSError errorWithDomain:@"openHAB" code:100 userInfo:errorDetail];
            [self.delegate openHABTrackingError:trackingError];
            self.reach = [Reachability reachabilityForInternetConnection];
            oldReachabilityStatus = [reach currentReachabilityStatus];
            [[NSNotificationCenter defaultCenter] addObserver: self
                                                     selector: @selector(reachabilityChanged:)
                                                         name: kReachabilityChangedNotification
                                                       object: nil];
            [self.reach startNotifier];
        }
    }
}

- (void)trackedLocalUrl
{
    if (self.delegate && [self.delegate respondsToSelector:@selector(openHABTrackingProgress:)]) {
        [self.delegate openHABTrackingProgress:@"Connecting to local URL"];
    }
    NSString *openHABUrl = [self normalizeUrl:openHABLocalUrl];
    [self trackedUrl:openHABUrl];
}

- (void)trackedRemoteUrl
{
    NSString *openHABUrl = [self normalizeUrl:openHABRemoteUrl];
    if ([openHABUrl length] > 0) {
        if (self.delegate && [self.delegate respondsToSelector:@selector(openHABTrackingProgress:)]) {
            [self.delegate openHABTrackingProgress:@"Connecting to remote URL"];
        }
        [self trackedUrl:openHABUrl];
    } else {
        if (self.delegate && [self.delegate respondsToSelector:@selector(openHABTrackingError:)]) {
            NSMutableDictionary *errorDetail = [NSMutableDictionary dictionary];
            [errorDetail setValue:@"Remote URL is not configured." forKey:NSLocalizedDescriptionKey];
            NSError *trackingError = [NSError errorWithDomain:@"openHAB" code:101 userInfo:errorDetail];
            [self.delegate openHABTrackingError:trackingError];
        }
    }
}

- (void)trackedDiscoveryUrl:(NSString *)discoveryUrl
{
    if (self.delegate && [self.delegate respondsToSelector:@selector(openHABTrackingProgress:)]) {
        [self.delegate openHABTrackingProgress:@"Connecting to discovered URL"];
    }
    [self trackedUrl:discoveryUrl];
}

- (void)trackedDemoMode
{
    if (self.delegate && [self.delegate respondsToSelector:@selector(openHABTrackingProgress:)]) {
        [self.delegate openHABTrackingProgress:@"Running in demo mode. Check settings to disable demo mode."];
    }
    [self trackedUrl:@"http://demo.openhab.org:8080"];
}

- (void)trackedUrl:(NSString *)trackedUrl
{
    if (self.delegate)
        [self.delegate openHABTracked:trackedUrl];
}

- (void) reachabilityChanged: (NSNotification *)notification {
    Reachability *changedReach = [notification object];
    if( [changedReach isKindOfClass: [Reachability class]]) {
        NetworkStatus nStatus = [changedReach currentReachabilityStatus];
        if (nStatus != oldReachabilityStatus) {
            NSLog(@"Network status changed from %@ to %@", [self stringFromStatus:oldReachabilityStatus], [self stringFromStatus:nStatus]);
            oldReachabilityStatus = nStatus;
            if (self.delegate && [self.delegate respondsToSelector:@selector(openHABTrackingNetworkChange:)]) {
                [self.delegate openHABTrackingNetworkChange:nStatus];
            }
        }
    }
}

- (void)startDiscovery
{
    NSLog(@"OpenHABTracking starting Bonjour discovery");
    if (self.delegate && [self.delegate respondsToSelector:@selector(openHABTrackingProgress:)]) {
        [self.delegate openHABTrackingProgress:@"Discovering openHAB"];
    }
    netService = [[NSNetService alloc] initWithDomain:@"local." type:@"_openhab-server-ssl._tcp." name:@"openHAB-ssl"];
    [netService setDelegate:self];
    [netService resolveWithTimeout:5.0];
}

// NSNetService delegate methods for Bonjour resolving
- (void)netServiceDidResolveAddress:(NSNetService *)resolvedNetService
{
    NSLog(@"OpenHABTracker discovered %@:%@", [self getStringIpFromAddressData:resolvedNetService.addresses[0]], [self getStringPortFromAddressData:resolvedNetService.addresses[0]]);
    NSString *openhabUrl = [NSString stringWithFormat:@"https://%@:%@", [self getStringIpFromAddressData:resolvedNetService.addresses[0]], [self getStringPortFromAddressData:resolvedNetService.addresses[0]]];
    [self trackedDiscoveryUrl:openhabUrl];
}

- (void)netService:(NSNetService *)netService
     didNotResolve:(NSDictionary *)errorDict
{
    NSLog(@"OpenHABTracker discovery didn't resolve openHAB");
    [self trackedRemoteUrl];
}

- (NSString *)normalizeUrl:(NSString *)url
{
    if ([url hasSuffix:@"/"]) {
        url = [url substringToIndex:url.length - 1];
    }
    return url;
}

- (BOOL) validateUrl: (NSString *) url {
    NSString *theURL =
    @"(http|https)://((\\w)*|([0-9]*)|([-|_])*)+([\\.|/]((\\w)*|([0-9]*)|([-|_])*))+";
    NSPredicate *urlTest = [NSPredicate predicateWithFormat:@"SELF MATCHES %@", theURL];
    return [urlTest evaluateWithObject:url];
}

- (BOOL)isNetworkConnected
{
    struct sockaddr_in zeroAddress;
    bzero(&zeroAddress, sizeof(zeroAddress));
    zeroAddress.sin_len = sizeof(zeroAddress);
    zeroAddress.sin_family = AF_INET;
    SCNetworkReachabilityRef defaultRouteReachability = SCNetworkReachabilityCreateWithAddress(NULL, (struct sockaddr *)&zeroAddress);
    SCNetworkReachabilityFlags flags;
    BOOL didRetrieveFlags = SCNetworkReachabilityGetFlags(defaultRouteReachability, &flags);
    CFRelease(defaultRouteReachability);
    if (!didRetrieveFlags)
    {
        return NO;
    }
    BOOL isReachable = flags & kSCNetworkFlagsReachable;
    BOOL needsConnection = flags & kSCNetworkFlagsConnectionRequired;
    return (isReachable && !needsConnection) ? YES : NO;
}

- (BOOL)isNetworkConnected2
{
    Reachability *networkReach = [Reachability reachabilityForInternetConnection];
    NetworkStatus networkReachabilityStatus = [networkReach currentReachabilityStatus];
    NSLog(@"Network status = %d", networkReachabilityStatus);
    if (networkReachabilityStatus == ReachableViaWiFi || networkReachabilityStatus == ReachableViaWWAN) {
        return YES;
    }
    return NO;
}

- (BOOL)isNetworkWiFi
{
    Reachability *wifiReach = [Reachability reachabilityForInternetConnection];
    NetworkStatus wifiReachabilityStatus = [wifiReach currentReachabilityStatus];
    if (wifiReachabilityStatus == ReachableViaWiFi) {
        return YES;
    }
    return NO;
}

- (NSString *)getStringIpFromAddressData:(NSData *)dataIn {
    struct sockaddr_in  *socketAddress = nil;
    NSString            *ipString = nil;
    
    socketAddress = (struct sockaddr_in *)[dataIn bytes];
    ipString = [NSString stringWithFormat: @"%s",
                inet_ntoa(socketAddress->sin_addr)];  ///problem here
    return ipString;
}

- (NSString *)getStringPortFromAddressData:(NSData *)dataIn {
    struct sockaddr_in  *socketAddress = nil;
    NSString            *ipPort = nil;
    
    socketAddress = (struct sockaddr_in *)[dataIn bytes];
    ipPort = [NSString stringWithFormat: @"%hu",
              ntohs(socketAddress->sin_port)];  ///problem here
    return ipPort;
}

- (BOOL)isURLReachable:(NSURL*) url
{
    FastSocket *client = [[FastSocket alloc] initWithHost:[url host] andPort:[[url port] stringValue]];
    NSLog(@"Checking if %@:%@ is reachable", [url host], [[url port] stringValue]);
    if ([client connect:1]) {
        [client close];
        return YES;
    }
    return NO;
}

- (NSString *)stringFromStatus:(NetworkStatus) status {
    
    NSString *string;
    switch(status) {
        case NotReachable:
            string = @"unreachable";
            break;
        case ReachableViaWiFi:
            string = @"WiFi";
            break;
        case ReachableViaWWAN:
            string = @"WWAN";
            break;
        default:
            string = @"Unknown";
            break;
    }
    return string;
}

@end

File: openHAB/OpenHABTracker.swift
// Copyright (c) 2010-2023 Contributors to the openHAB project
//
// See the NOTICE file(s) distributed with this work for additional
// information.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0
//
// SPDX-License-Identifier: EPL-2.0

import Alamofire
import Foundation
import OpenHABCore
import os.log
import SystemConfiguration

protocol OpenHABTrackerDelegate: AnyObject {
    func openHABTracked(_ openHABUrl: URL?, version: Int)
    func openHABTrackingProgress(_ message: String?)
    func openHABTrackingError(_ error: Error)
}

protocol OpenHABTrackerExtendedDelegate: OpenHABTrackerDelegate {
    func openHABTrackingNetworkChange(_ networkStatus: NetworkReachabilityManager.NetworkReachabilityStatus)
}

class OpenHABTracker: NSObject {
    static var shared = OpenHABTracker()

    public var multicastDelegate = MulticastDelegate<OpenHABTrackerDelegate>()
    private var oldReachabilityStatus: NetworkReachabilityManager.NetworkReachabilityStatus?
    private let reach = NetworkReachabilityManager()
    private var openHABDemoMode = false
    private var openHABLocalUrl = ""
    private var openHABRemoteUrl = ""
    private var netService: NetService?
    private var restartTimer: Timer?

    var appData: OpenHABDataObject? {
        AppDelegate.appDelegate.appData
    }

    override private init() {
        super.init()
        NotificationCenter.default.addObserver(self, selector: #selector(restart), name: NSNotification.Name("org.openhab.preferences.saved"), object: nil)
        start()
    }

    @objc
    func restart() {
        reach?.stopListening()
        start()
    }

    private func start() {
        openHABDemoMode = Preferences.demomode
        openHABLocalUrl = Preferences.localUrl
        openHABRemoteUrl = Preferences.remoteUrl

        #if DEBUG
        // always activate demo mode for UITest
        if ProcessInfo.processInfo.environment["UITest"] != nil {
            openHABDemoMode = true
        }
        #endif

        // Start NetworkReachabilityManager.Listener
        oldReachabilityStatus = reach?.status
        reach?.startListening { [weak self] status in
            guard let self else { return }
            let nStatus = status
            // use a timer to prevent bouncing/flapping around when switching between wifi, vpn, and wwan
            self.restartTimer?.invalidate()
            self.restartTimer = Timer.scheduledTimer(withTimeInterval: 3.0, repeats: false) { _ in
                if nStatus != self.oldReachabilityStatus {
                    if let oldReachabilityStatus = self.oldReachabilityStatus {
                        os_log("OpenHABTracker Network status changed from %{PUBLIC}@ to %{PUBLIC}@", log: OSLog.remoteAccess, type: .info, self.string(from: oldReachabilityStatus) ?? "", self.string(from: nStatus) ?? "")
                    }
                    self.oldReachabilityStatus = nStatus
                    if self.isNetworkConnected() {
                        self.restart()
                    }
                }
            }
        }

        // Check if any network is available
        if isNetworkConnected() {
            // Check if demo mode is switched on in preferences
            if openHABDemoMode {
                os_log("OpenHABTracker demo mode preference is on", log: .default, type: .info)
                tryDemoMode()
            } else {
                if isNetworkWiFi(), openHABLocalUrl.isEmpty {
                    startDiscovery()
                } else {
                    os_log("OpenHABTracker network trying all", log: .default, type: .info)
                    tryAll()
                }
            }
        } else {
            os_log("OpenHABTracker network not available", log: .default, type: .info)
            multicastDelegate.invoke { $0.openHABTrackingError(errorMessage("network_not_available")) }
        }
    }

    private func tryDiscoveryUrl(_ discoveryUrl: URL?) {
        multicastDelegate.invoke { $0.openHABTrackingProgress(NSLocalizedString("connecting_discovered", comment: "")) }
        tryUrl(discoveryUrl)
    }

    private func tryDemoMode() {
        multicastDelegate.invoke { $0.openHABTrackingProgress(NSLocalizedString("running_demo_mode", comment: "")) }
        tryUrl(URL(staticString: "https://demo.openhab.org"))
    }

    /// Attemps to connect to the URL and get the openHAB version
    /// - Parameter tryUrl: Completes with the url and version of openHAB that succeeded, or an Error object if failed
    private func tryUrl(_ tryUrl: URL?) {
        getServerInfoForUrl(tryUrl) { url, version, error in
            if let error {
                self.multicastDelegate.invoke { $0.openHABTrackingError(error) }
            } else {
                self.appData?.openHABVersion = version
                self.appData?.openHABRootUrl = url?.absoluteString ?? ""
                self.multicastDelegate.invoke { $0.openHABTracked(url, version: version) }
            }
        }
    }

    /// Attemps to connect in parallel to the remote and local URLs if configured, the first URL to succeed wins
    private func tryAll() {
        var urls = [(url: String, delay: Double)]()
        if !openHABLocalUrl.isEmpty {
            urls.append((url: openHABLocalUrl, delay: 0.0))
        }
        if !openHABRemoteUrl.isEmpty {
            urls.append((url: openHABRemoteUrl, delay: openHABLocalUrl.isEmpty ? 0 : 1.5))
        }
        if urls.isEmpty {
            multicastDelegate.invoke { $0.openHABTrackingError(errorMessage("error")) }
            return
        }
        multicastDelegate.invoke { $0.openHABTrackingProgress(NSLocalizedString("connecting", comment: "")) }
        tryUrls(urls) { url, version, error in
            if let error {
                os_log("OpenHABTracker failed %{PUBLIC}@", log: .default, type: .info, error.localizedDescription)
                self.multicastDelegate.invoke { $0.openHABTrackingError(error) }
            } else {
                self.appData?.openHABVersion = version
                self.appData?.openHABRootUrl = url?.absoluteString ?? ""
                self.multicastDelegate.invoke { $0.openHABTracked(url, version: version) }
            }
        }
    }

    /// Tries to connect in parallel to all URL's passed in and completes when either the first requests succeedes, or all fail.
    /// - Parameters:
    ///   - urls: Tuple of String URLS and a request Delay value
    ///   - completion: Completes with the url and version of openHAB that succeeded, or an Error object if all failed
    private func tryUrls(_ urls: [(url: String, delay: Double)], completion: @escaping (URL?, Int, Error?) -> Void) {
        var isRequestCompletedSuccessfully = false
        // request in flight
        var requests = [DataRequest]()
        // timers that have yet to be executed
        var timers = [(url: URL, timer: Timer)]()
        for item in urls {
            let url = URL(string: item.url)!
            let restUrl = URL(string: "rest/", relativeTo: url)!
            let timer = Timer.scheduledTimer(withTimeInterval: item.delay, repeats: false) { _ in
                let request = NetworkConnection.shared.manager.request(restUrl, method: .get)
                    .validate()
                requests.append(request)
                // remove us from the outstanding timer list
                timers.removeAll(where: { $0.url == url })
                request.responseData { response in
                    // remove us from the outstanding request list
                    requests.removeAll(where: { $0 == request })
                    switch response.result {
                    case let .success(data):
                        let version = self.getServerInfoFromData(data: data)
                        if version > 0, !isRequestCompletedSuccessfully {
                            isRequestCompletedSuccessfully = true
                            completion(url, version, nil)
                            // cancel any timers not yet fired
                            timers.forEach { $0.timer.invalidate() }
                            // cancel any requests still in flight
                            requests.forEach { $0.cancel() }
                        }
                    case let .failure(error):
                        os_log("OpenHABTracker request failure %{PUBLIC}@", log: .notifications, type: .error, error.localizedDescription)
                    }
                    // check if we are the last attempt
                    if !isRequestCompletedSuccessfully, requests.isEmpty, timers.isEmpty {
                        os_log("OpenHABTracker last response", log: .notifications, type: .error)
                        if !isRequestCompletedSuccessfully {
                            completion(nil, 0, self.errorMessage("network_not_available"))
                        }
                    }
                }
                request.resume()
            }
            timers.append((url, timer))
            RunLoop.main.add(timer, forMode: .common)
        }
    }

    /// Attempts to parse the data response from a request and determine if its an openHAB server and its server version
    /// - Parameter data: request data
    /// - Returns: Version of openHAB or -1 if not an openHAB server
    private func getServerInfoFromData(data: Data) -> Int {
        do {
            let serverProperties = try data.decoded(as: OpenHABServerProperties.self)
            os_log("OpenHABTracker openHAB version %@", log: .remoteAccess, type: .info, serverProperties.version)
            // OH versions 2.0 through 2.4 return "1" as thier version, so set the floor to 2 so we do not think this is a OH 1.x serevr
            return max(2, Int(serverProperties.version) ?? 2)
        } catch {
            // testing for OH 1.x
            let str = String(decoding: data, as: UTF8.self)
            if str.hasPrefix("<?xml") {
                return 1
            } else {
                os_log("OpenHABTracker Could not decode response", log: .notifications, type: .error)
                return -1
            }
        }
    }

    /// Attempts to connect to a URL and determine its server version
    /// - Parameters:
    ///   - url: URL of the openHAB server
    ///   - completion: Completes with the url and version of openHAB that succeeded, or an Error object if failed
    private func getServerInfoForUrl(_ url: URL?, completion: @escaping (URL?, Int, Error?) -> Void) {
        let strUrl = url?.absoluteString ?? ""
        os_log("OpenHABTracker getServerInfo, trying: %{PUBLIC}@", log: .default, type: .info, strUrl)
        NetworkConnection.tracker(openHABRootUrl: strUrl) { response in
            os_log("OpenHABTracker getServerInfo, recieved data for URL: %{PUBLIC}@", log: .default, type: .info, strUrl)
            switch response.result {
            case let .success(data):
                let version = self.getServerInfoFromData(data: data)
                if version > 0 {
                    completion(url, version, nil)
                } else {
                    completion(url, 0, self.errorMessage("error"))
                }
            case let .failure(error):
                os_log("OpenHABTracker getServerInfo ERROR for %{PUBLIC}@ : %{PUBLIC}@ %d", log: .remoteAccess, type: .error, strUrl, error.localizedDescription, response.response?.statusCode ?? 0)
                completion(url, 0, error)
            }
        }
    }

    private func startDiscovery() {
        os_log("OpenHABTracking starting Bonjour discovery", log: .default, type: .info)
        multicastDelegate.invoke { $0.openHABTrackingProgress(NSLocalizedString("discovering_oh", comment: "")) }
        netService = NetService(domain: "local.", type: "_openhab-server-ssl._tcp.", name: "openHAB-ssl")
        netService!.delegate = self
        netService!.resolve(withTimeout: 5.0)
    }

    func normalizeUrl(_ url: String?) -> String? {
        if let url, url.hasSuffix("/") {
            return String(url.dropLast())
        }
        return url
    }

    func isNetworkConnected() -> Bool {
        reach?.isReachable ?? false
    }

    func isNetworkWiFi() -> Bool {
        reach?.isReachableOnEthernetOrWiFi ?? false
    }

    func string(from status: NetworkReachabilityManager.NetworkReachabilityStatus) -> String? {
        switch status {
        case .unknown, .notReachable:
            return "unreachable"
        case let .reachable(connectionType):
            return connectionType == .ethernetOrWiFi ? "WiFi" : "WWAN"
        }
    }

    func errorMessage(_ message: String) -> NSError {
        var errorDetail: [AnyHashable: Any] = [:]
        errorDetail[NSLocalizedDescriptionKey] = NSLocalizedString(message, comment: "")
        return NSError(domain: "openHAB", code: 101, userInfo: errorDetail as? [String: Any])
    }
}

extension OpenHABTracker: NetServiceDelegate, NetServiceBrowserDelegate {
    // NSNetService delegate methods for publication
    func netServiceDidResolveAddress(_ resolvedNetService: NetService) {
        func getStringIp(fromAddressData dataIn: Data?) -> String? {
            var ipString: String?
            let data = dataIn! as NSData
            let socketAddress: sockaddr_in = data.castToCPointer()
            ipString = String(cString: inet_ntoa(socketAddress.sin_addr), encoding: .ascii)
            return ipString
        }

        guard let data = resolvedNetService.addresses?.first else { return }
        let resolvedComponents: URLComponents = {
            var components = URLComponents()
            components.host = getStringIp(fromAddressData: data)
            components.scheme = "https"
            components.port = resolvedNetService.port
            return components
        }()

        let openhabUrl = "\(resolvedComponents.url!)"
        os_log("OpenHABTracker discovered:%{PUBLIC}@ ", log: OSLog.remoteAccess, type: .info, openhabUrl)
        tryDiscoveryUrl(resolvedComponents.url)
    }

    func netService(_ netService: NetService, didNotResolve errorDict: [String: NSNumber]) {
        os_log("OpenHABTracker discovery didn't resolve openHAB", log: .default, type: .info)
        tryAll()
    }
}

extension NSData {
    func castToCPointer<T>() -> T {
        let mem = UnsafeMutablePointer<T>.allocate(capacity: MemoryLayout<T.Type>.size)
        getBytes(mem, length: MemoryLayout<T.Type>.size)
        return mem.move()
    }
}
